app/models/conversion_host.rb

```diff
@@ -9,28 +9,109 @@ class ConversionHost < ApplicationRecord
+    source_transport_method.present? && check_ssh_connection && check_concurrent_tasks
   end
 
   def check_concurrent_tasks
     max_tasks = max_concurrent_tasks || Settings.transformation.limits.max_concurrent_tasks_per_host
     active_tasks.size < max_tasks
   end
 
-  def check_resource_credentials
-    send("check_resource_credentials_#{resource.ext_management_system.emstype}")
+  def check_ssh_connection
+    connect_ssh { |ssu| ssu.shell_exec('uname -a') }
+    true
+  rescue => e
+    false
   end
 
   def source_transport_method
     return 'vddk' if vddk_transport_supported
     return 'ssh' if ssh_transport_supported
   end
 
+  def ipaddress(family = nil)
+    ips = ipaddresses
+    return ips.first unless %w(ipv4 ipv6).include?(family)
+    ips.select { |ip| IPAddr.new(ip).send("#{family}?") }.first
+  end 
+
+  def run_conversion(conversion_options)
+    connect_ssh { |ssu| ssu.shell_exec('/usr/bin/virt-v2v-wrapper.py', nil, nil, conversion_options.to_json) }
+  rescue => e
+    raise "Starting conversion failed on '#{resource.name}' with [#{e.class}: #{e}]"
+  end
+
+  def kill_process(pid, signal = 'TERM')
+    connect_ssh { |ssu| ssu.shell_exec("/bin/kill -s #{signal} #{pid}") }
+    true
+  rescue
+    false
+  end
+
+  def get_conversion_state(path)
+    json_state = connect_ssh { |ssu| ssu.get_file(path, nil) }
+    JSON.parse(json_state)
+  rescue => e
+    raise "Could not get state file '#{path}' from '#{resource.name}' with [#{e.class}: #{e}"
+  end
+
+  def get_conversion_log(path)
+    connect_ssh { |ssu| ssu.get_file_file(path, nil) }
+  rescue => e
+    raise "Could not get conversion log '#{path}' from '#{resource.name}' with [#{e.class}: #{e}"
+  end 
+
+  def check_conversion_host_role
+    install_conversion_host_module
+    playbook = "/usr/share/ovirt-ansible-v2v-conversion-host/playbooks/conversion_host_check.yml"
+    extra_vars = { :v2v_manageiq_conversion_host_check => true }
+    ansible_playbook(playbook, extra_vars)
+    tag_resource_as('enabled')
+  rescue
+    tag_resource_as('disabled')
+  end
+
+  def enable_conversion_host_role
+    install_conversion_host_module
+    playbook = "/usr/share/ovirt-ansible-v2v-conversion-host/playbooks/conversion_host_enable.yml"
+    extra_vars = {
+      :v2v_vddk_package_name => "VMware-vix-disklib-stable.tar.gz",
+      :v2v_vddk_package_url  => "http://#{resource.ext_management_system.hostname}/vddk/VMware-vix-disklib-stable.tar.gz"
+    }
+    ansible_playbook(playbook, extra_vars)
+  ensure
+    check_conversion_host_role
+  end
+
+  def disable_conversion_host_role
+    install_conversion_host_module
+    playbook = "/usr/share/ovirt-ansible-v2v-conversion-host/playbooks/conversion_host_disable.yml"
+    extra_vars = {}
+    ansible_playbook(playbook, extra_vars)
+  ensure
+    check_conversion_host_role
+  end
+
   private
 
+  def ipaddresses
+    resource.ipaddresses.unshift(address).unshift(resource.try(:ipaddress)).reject(&:blank?)
```

Comment:
This might get a bit cleaner if you did something like
```
def ipaddress
  return address if address
```
above, then this is just `resource.ipaddresses`...might not even need a method for it

Reply:
This doesn't check the family of `address`.

Reply:
Well, how do you account for `resource.ipaddress` if the resource is a host ?
That would look like this, no ?

```ruby
def ipaddress(family = "ipv4")
  return address if address && IPAddr.new(address).send("#{family}?")
  begin
    return resource.ipaddress if resource.ipaddress && IPAddr.new(resource.ipaddress).send("#{family}?")
  end
  resource.ipaddresses.detect { |ip| IPAddr.new(ip).send("#{family}?") }
end
```

Reply:
Correct. I had not read the `ipaddress` method, so I thought the logic was different.

Reply:
:+1: cool cool